modules/docs/src/05-design/03-services.md

```diff
@@ -0,0 +1,293 @@
+
+* Finally-encoded algebras are object-oriented encodings of a set of operations, just like above : operations are represented as **methods** in an interface. Interpretation of expressions written in terms of these methods does not involve any runtime transformation from one context to another : the method call is merely executed. In other words, when they are executed, expressions coming from finally-encoded algebras are already in their "final form".
+
+* Conversely, initially-encoded algebras represent expressions as **data**, implying that interpretation involves a transformation of this data into lower level method calls. However, **data** has the quality of being a first class construct in programming languages, meaning you can pass it around and use it as parameter to functions. This allows for the unification of codepaths, as the differences between some aspects of a bit of logic can be absorbed by the data and handled later on.
+
+Finally-encoded KVStore algebra :
+
+```scala
+trait KVStore[Context[_]]{
+  def put(key: String, value: String) : Context[Unit]
+  def get(key: String)                : Context[Option[String]]
+  def delete(key: String)             : Context[Unit]
+}
+```
+
+Initially-encoded KVStore algebra :
+
+```scala
+sealed trait KVStoreOp[Output]
+object KVStoreOp {
+  case class Put(key: String, value: String)  extends KVStore[Unit]
+  case class Get(key: String)                 extends KVStore[Option[String]]
+  case class Delete(key: String)              extends KVStore[Unit]
+}
+```
+
+These two encodings contain a similar amount of information. It is even nearly-trivial to go from a `KVstore[Context]` instance to a `KVStoreOp ~> Context` natural transformation, and vice versa:
+
+```scala
+trait ~>[F[_], G[_]]{
+  def apply[A](fa: F[A]) : G[A]
+}
+
+def asNaturalTransformation[Context[_]](impl: KVStore[Context]) = new (KVStoreOp ~> Context){
+  def apply[A](fa: KVStoreOp[A]) : Context[A] = fa match {
+    case KVStoreOp.Put(key, value) => impl.put(key, value)
+    case KVStoreOp.Get(key)        => impl.get(key)
+    case KVStoreOp.Delete(key)     => impl.delete(keyu)
+  }
+}
+
+def fromNaturalTransformation[Context[_]](run: KVStoreOp ~> Context) = new KVStore[Context]{
+  def put(key: String, value: String) = run(KVStoreOp.Put(key, value))
+  def get(key: String)                = run(KVStoreOp.Get(key))
+  def delete(key: String)             = run(KVStoreOp.Delete(key))
+}
+```
+
+This duality is heavily used by Smithy4s : finally-encoded interfaces are generally more natural to Scala developers, and are better supported in editors (autocompletion, etc). But from an implementation's perspective, the initial, data-base encoding is really interesting, because operations are reified as **datatypes** that can be associated with instances of generic typeclasses : it is possible to abstract over data, it is not possible to abstract over method calls.
```

Comment:
```suggestion
This duality is heavily used by Smithy4s : finally-encoded interfaces are generally more natural to Scala developers, and are better supported in editors (autocompletion, etc). But from an implementation's perspective, the initial, data-based encoding is really interesting, because operations are reified as **datatypes** that can be associated with instances of generic typeclasses : it is possible to abstract over data, it is not possible to abstract over method calls.
```